lib/Infer/ExhaustiveSynthesis.cpp

```diff
@@ -127,20 +132,41 @@ void addGuess(Inst *RHS, int MaxCost, std::vector<Inst *> &Guesses,
     TooExpensive++;
 }
 
-// most naive prune algorithm ever
-bool prune (Inst *I, std::vector<Inst *> &ReservedInsts) {
-  if (instCount(I) >= MaxNumInstructions)
-    return false;
-  if (ReservedInsts.empty())
+typedef std::function<bool(Inst *, std::vector<Inst *> &)> PruneFunc;
+
+// Does a short-circuiting AND operation
+PruneFunc MkPruneFunc(std::vector<PruneFunc> Funcs) {
+  return [Funcs](Inst *I, std::vector<Inst *> &RI) {
+    for (auto F : Funcs) {
+      if (!F(I, RI)) {
+        return false;
+      }
+    }
+    return true;
+  };
+}
+
+// return false if successfully proved infeasible
+bool CostPrune(Inst *I, std::vector<Inst *> &ReservedInsts) {
+
+  // Cost exceeds LHS
+  if (!ReservedInsts.empty() && instCount(I) >= MaxNumInstructions)
     return false;
+
+//   if (ReservedInsts.empty() && instCount(I) > MaxNumInstructions)
+//     return false;
+// TODO : Handle this logic here instead of comparing against TooExpensive
+//  at arbitrary places in the synthesis algorithm
+
   return true;
 }
 
 void getGuesses(std::vector<Inst *> &Guesses,
                 const std::vector<Inst *> &Inputs,
                 int Width, int LHSCost,
                 InstContext &IC, Inst *PrevInst, Inst *PrevSlot,
-                int &TooExpensive) {
+                int &TooExpensive,
+                PruneFunc prune) {
```

Comment:
locals should start with a capital (only marking this example but there are others you should fix)

Reply:
Fixed